### ISSUE TITLE
Implementation and application of material change using pymoab

### ISSUE THREAD

**Issue by zxkjack123** [pyne/pyne#1163]
This issue comes from the conversation between @gonuke @pshriwise and @zxkjack123  in PR #1159.
Here is the short summary of the topic "implementation and application of material change using pymoab":
Implementation
  - In Trelis:
      - Volumes whose materials will not change are assigned as usual in Trelis for neutron transport (`mat:steel/rho:8.08`) and applied as normal
      - Volumes whose material change in the photon transport step will have a different material name,  something like `mat:steel/rho:8.08|<application>_mat:aluminum/rho:2.7`
  - In PyMOAB:
      - Add function to change the material of:
          - Volume: Change volume material, we can use it in the case of only volume material change. Eg. R2S photon transport, material sensitivity analysis, moving volume
          - Mesh volume element: Change the material of mesh volume element. We can use it to automatically update the material of the mesh element. We can use it in the case related to time dependent (depletion/transmutation considered) analysis or moving mesh element.
  - In DAGMC/UWUW:
    - A material assignment checking.
      - Make sure every volume has a material;
      - And no volume has more than three materials.  (If there is no conflict when we assign material in Trelis, that means we may assignment two different neutron or photon material for a volume)
      - If a volume is in two groups, they must be a `mat` or `<application>_mat`
    - Add `<application>_mat` support. At least we need to let the code know that we want to store these materials into the geometry.

Applications:
- Different running mode with different material. For example, in R2S, use different material for the same volume to represent the component change/remove/insection.
- Time step changes geometry. Such as time step R2S of moving system that Chelsea has studied.
- Some (groups of) volumes have different material setting in different cases for material sensitivity analysis.
- Material updates considering the depletion/transmutation: coupling with depletion/transmutation codes, and with the help of PyMOAB to change the material, a time dependent simulation could be achieved.

To do:
- [ ] : Discuss the identifing name for each application. Or could we use the uniformly new name such as `new_mat`? As we have different applications.
- [ ] : Discuss the implementation and interfaces. Change the material of volume related to Trelis material assignment. But change the material of a mesh element is slightly different.
- [ ] : The user input part. When, where and how should we provide the new materials?
- [ ] : More ...

**Comment by zxkjack123** [comment]
@gonuke @pshriwise I summarized the current state of the conversation as above. Is there something I am wrong or omit?

**Comment by pshriwise** [comment]
I would describe the changes in PyMOAB more as changes in PyNE (or any other application leveraging this material assignment scheme). 

#### Regarding material names and DagMC/UWUW modifications

 My personal preference is for all materials to have a "default" name like `mat:steel/rho:8.08`. In addition, the volumes whose materials need to be updated can be added to another group `<application>_mat:7/rho:2.7`. When the application recognizes that the simulation is a specific photon transport run/time step/depletion step, the volume's additional names will be searched. If an appropriate name is found for the volume, it's material assignment will be updated *prior to the simulation initialization using PyMOAB* - meaning that we will read in the original `.h5m`, modify it, and write out a new .h5m file and then run the simulation with the new `.h5m`. This method should require no changes to DagMC/UWUW because material assignments will always be valid. 

I prefer this method because it keeps application-specific code out of DagMC/UWUW, keeping our conditions for valid material assignments simple there.

**Comment by zxkjack123** [comment]
@pshriwise I agree to use `<application>_mat` format. Because I find there is a possiblity to put more than 1 applications in one geometry. We can assign `<application1>_mat` and `<application2>_mat` in the same geometry, and chose which to replace according to the specific application.
